### ISSUE TITLE
Create simple Bluesky acquisition plan independent from DM calls

### ISSUE THREAD

**Issue by qzhang234** [aps-8id-dys/bluesky#95]
8-ID-I would like to have a simple Bluesky acquisition plan that does only the following:

1. Start the cam module to acquire;
2. Start the HDF plugin to save images.

We would prefer that this Bluesky plan DOES NOT READ/WRITE ANY PV unless specifically asked. This includes staging/unstaging. The beamline scientists will be responsible for setting the correct detector configuration, which includes providing the file name and file path to the HDF plugin. The beamline scientists will also be responsible for any sample motion/temperature control that is separate from the detector acquisition.

In addition, given that the x-ray measurement should not wait on the completion of DM to proceed, we would like to completely separate the DM part from the Bluesky acquisition. This will reduce the complexity and improve the stability of the code, and also makes it easier to debug and maintain.

Please let me know what your thoughts are on this.

**Comment by prjemian** [comment]
The code for that plan exists within the current full plan: https://github.com/aps-8id-dys/bluesky/blob/d42eb4d72fbb659d184df8406408c0ab0de53d1f/src/instrument/plans/bdp_demo.py#L356-L380

**Comment by qzhang234** [comment]
That looks very nice. I have two questions regarding that function:

1. Is it absolutely necessary to return the uuid of `bp.count`?
2. Can `bp.count` work with an empty `stage_sigs`?

**Comment by prjemian** [comment]
@qzhang234  Should this plan accept arguments for the acquisition parameters?  (Lines 367-373).

> 1. Is it absolutely necessary to return the uuid of `bp.count`?

Yes.

**Comment by qzhang234** [comment]
> @qzhang234 Should this plan accept arguments for the acquisition parameters? (Lines 367-373).
> 
> > 1. Is it absolutely necessary to return the uuid of `bp.count`?
> 
> Yes.

@prjemian I think just the `bp.count` with HDF plugin should be fine for now. I'll wrap it with another function where I pass all the variables manually. Would this be possible?

**Comment by prjemian** [comment]
Looking like this in my editor.

![image](https://github.com/user-attachments/assets/b9ef00e8-06c2-4fd2-ace8-7aee54f7977f)

**Comment by prjemian** [comment]
What about the `nxwriter` here?

**Comment by prjemian** [comment]
If the `nxwriter` is dropped, the plan reduces to `bp.count([det], md=md)`

**Comment by qzhang234** [comment]
> What about the `nxwriter` here?

Can the entire `nxwriter` be made separate from the acquisition? Also is there a reason that we would need `md=md` inside `bp.count`?

**Comment by qzhang234** [comment]
I think it's better to separate the `nxwriter` from `bp.count`, because we want Bluesky to proceed to the next measurement even if the metadata is entered incorrectly

**Comment by prjemian** [comment]
You want `bp.count([det], md=md)` (rather than `bp.count([det])`) so the user can add metadata to this plan.  You don't want to block that.  The `DEFAULT_RUN_METADATA` is a suggestion so the Python style checker does not complain about an empty dictionary.  This is the complaint:

```py
src/instrument/plans/independent.py:23:16: B006 Do not use mutable data structures for argument defaults
   |
21 |     # num_triggers=num_triggers,
22 |     # path=data_path,
23 |     md: dict = {},
   |                ^^ B006
24 | ):
25 |     """
   |
   = help: Replace with `None`; initialize within function
```

A suggested dictionary is better than `None` which provides **no** suggestion to the user about what to provide.

**Comment by prjemian** [comment]
> all we need to do is to test if bp.count will work with empty stage_sigs, right?

Yes.  That is the simplest case.

**Comment by qzhang234** [comment]
> You want `bp.count([det], md=md)` (rather than `bp.count([det])`) so the user can add metadata to this plan. You don't want to block that. The `DEFAULT_RUN_METADATA` is a suggestion so the Python style checker does not complain about an empty dictionary. This is the complaint:
> 
> ```python
> src/instrument/plans/independent.py:23:16: B006 Do not use mutable data structures for argument defaults
>    |
> 21 |     # num_triggers=num_triggers,
> 22 |     # path=data_path,
> 23 |     md: dict = {},
>    |                ^^ B006
> 24 | ):
> 25 |     """
>    |
>    = help: Replace with `None`; initialize within function
> ```
> 
> A suggested dictionary is better than `None` which provides **no** suggestion to the user about what to provide.

But I thought we don't need to add metadata to `bp.count` for it to function? Everything such as the file name, acquisition time, etc. will be populated to EPICS IOC via `bps.mv` and then manually entered into the metadata file via `nxwriter`

**Comment by prjemian** [comment]
Correct.  You do not **need** to add metadata.  Similarly, you do not want to **prevent** a caller from supplying metadata.

You intend this plan to be called from another.  This plan must not break the chain by dropping any metadata provided by the caller.

**Comment by qzhang234** [comment]
Hi @prjemian, here's the pseudo code I have in mind, please see if this strategy will work:

![image](https://github.com/user-attachments/assets/fb9fbdfc-f7c6-4d19-9666-8b9b3e7cf401)
![image](https://github.com/user-attachments/assets/834ddc77-1ace-46bd-889a-a04575580446)

**Comment by prjemian** [comment]
uh, no.  There is no part of this that collects the data (`bp.count()`).

Would you like me to work that up for you in the branch?

The `write_nxfile()` plan does not write, it waits for the file to finish flushing to disk after the detector is complete.  You need it right after `bp.count()` so that the detector buffers are not cleared before the write finishes.

The `md=md` kwargs in the plan stubs is not necessaary.  It's only used by plans that (eventually) call `bps.open_run`.  This includes the plans from `bluesky.plans`.

Is the code above missing a separate plan for the JSON file (lines 18-22), just to keep it modular?

**Comment by qzhang234** [comment]
You are right about the JSON file part. Also I do not know what function writes the nx file so I was just guessing. 

However, could you create a plan like above with separate functions that each does its own thing, and keep the structure as flat as possible (e.g., no decorators and no new class definition)?

If you could dumb it down to a level that even someone like me can understand, then hopefully I'll be able to maintain it myself without having to bug you constantly. Also please keep each operation in its own separate functions so that it is easier to debug.

Thank you!

**Comment by prjemian** [comment]
Found it on line 10.  Needs to have nxwriter subscribed.

**Comment by prjemian** [comment]
Without the decorator for nxwriter, it gets very sloppy.  Users will always get this wrong.

**Comment by prjemian** [comment]
I'll post shortly and we can hack from there.  Better than screen views.

**Comment by prjemian** [comment]
We'll need to capture the bluesky run uid so we can exchange metadata with APS Data Management.  The DM workflow kickoff looks complicated because of all the metadata.

**Comment by qzhang234** [comment]
> Without the decorator for nxwriter, it gets very sloppy. Users will always get this wrong.

Hi @prjemian, one thing I would like to clarify is, users will never touch this code and neither will we expose it to the users. This code will only be used by us, to build customized functions for users. The only thing users are allowed to do is to change the input variables. 

For this reason, we want the bottom layer code to be as simple and as flat as it can be.

Would it be possible?

**Comment by prjemian** [comment]
All the new code is in `independent.py`.  Still some more things to do in the `kickoff_dm_workflow()` plan.

`full_acquisition()` is that bottom-layer code you described.

**Comment by qzhang234** [comment]
Hi @prjemian , this looks very nice!

I have one question: why do we need the decorator?
https://github.com/aps-8id-dys/bluesky/blob/4356248edcd654711fec9f0d03eb673bed85441d/src/instrument/plans/independent.py#L32-L34

**Comment by qzhang234** [comment]
Another question: how do we explicitly set up empty `stage_sigs` so that we are assured that Bluesky is not changing any other detector PVs during the acquisition?

**Comment by prjemian** [comment]
Let's do that tomorrow.  Simplest is `eiger4M.stage_sigs = {}`.  Since it's an area detector, nothing is quite as easy as that.  We've got some more investigating and work to do.

**Comment by prjemian** [comment]
The decorator subscribes the `nxwriter` to receive content for just this one data acquisition.  Since that's the metadata file, we don't want content from any other bluesky runs in the same file.  The decorator gets that job done.
